Sources/HTTP/SimpleSocket/SimpleServerSocket.swift

```diff
@@ -0,0 +1,152 @@
+// This source file is part of the Swift.org Server APIs open source project
+//
+// Copyright (c) 2017 Swift Server API project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+//
+
+import Foundation
+import Dispatch
+
+internal class SimpleServerSocket {
+    internal var socketfd: Int32 = -1
+    
+    internal var listeningPort: Int32 = -1
+    
+    internal private(set) var isListening = false
+    
+    internal private(set) var isConnected = false
+
+    internal func socketRead(into readBuffer: inout UnsafeMutablePointer<Int8>, maxLength:Int) throws -> Int {
+        //let readBuffer: UnsafeMutablePointer<CChar> = UnsafeMutablePointer<CChar>.allocate(capacity: maxLength)
+        readBuffer.initialize(to: 0x0)
+        let read = recv(self.socketfd, readBuffer, maxLength, Int32(0))
+//        if read > 0 {
+//            data.append(readBuffer, length: read)
+//        }
+        return read
+    }
+    
+    @discardableResult internal func socketWrite(from buffer: UnsafeRawPointer, bufSize: Int) throws -> Int {
+        return send(self.socketfd, buffer, Int(bufSize), Int32(0))
+    }
+    
+    internal func shutdownAndClose() {
+        //print("Shutting down socket \(self.socketfd)")
+        if self.isListening || self.isConnected {
+            //print("Shutting down socket")
+            _ = shutdown(self.socketfd, Int32(SHUT_RDWR))
+            self.isListening = false
+        }
+        self.isConnected = false
+        close(self.socketfd)
+    }
+        
+    internal func acceptClientConnection() throws -> SimpleServerSocket {
+        let retVal = SimpleServerSocket()
+        
+        var acceptFD: Int32 = -1
+        repeat {
+            var acceptAddr = sockaddr_in()
+            var addrSize = socklen_t(MemoryLayout<sockaddr_in>.size)
+            
+            acceptFD = withUnsafeMutablePointer(to: &acceptAddr) { pointer in
+                return accept(self.socketfd, UnsafeMutableRawPointer(pointer).assumingMemoryBound(to: sockaddr.self), &addrSize)
+            }
+            if acceptFD < 0 && errno != EINTR {
+                //fail
+                print("errno is \(errno)")
+            }
+            /*
+            else {
+                print("Accept returned fd \(acceptFD)")
+            }
+             */
+        }
+        while acceptFD < 0
+        
+        retVal.isConnected = true
+        retVal.socketfd = acceptFD
+        
+        return retVal
+    }
+    
+    internal func bindAndListen(on port: Int, maxBacklogSize: Int32 = 10000) throws {
+	#if os(Linux)
+        self.socketfd = socket(Int32(AF_INET), Int32(SOCK_STREAM.rawValue), Int32(IPPROTO_TCP))
```

Comment:
we set some internal constants in vapor to simplify areas like this https://github.com/vapor/vapor/blob/beta/Sources/TCP/Utilties/Utilities.swift#L28-L34

Reply:
Oh, wow. I love the touchup of `SOCK_STREAM`. @ianpartridge, is it okay with you if I mimic https://github.com/vapor/vapor/blob/beta/Sources/TCP/Utilties/Utilities.swift#L32 ?

Reply:
Sure, anything that makes things more readable is good.